tensorflow_privacy/privacy/membership_inference_attack/data_structures.py

```diff
@@ -444,6 +444,69 @@ def __str__(self):
     ])
 
 
+@dataclass
+class SingleRiskScoreResult:
+  """Results from computing privacy risk scores.
+  this part shows how to leverage privacy risk score to perform attacks with thresholding on risk score
+  """
+
+  # Data slice this result was calculated for.
+  slice_spec: SingleSliceSpec
+
+  train_risk_scores: np.ndarray 
+  
+  test_risk_scores: np.ndarray
+  
+  def attack_with_varied_thresholds(self, threshold_list):
+    """ For each threshold value, we count how many training and test samples with privacy risk scores larger than the threshold
+    and further compute precision and recall values.
+    We skip the threshold value if it is larger than every sample's privacy risk score.
+    """
+    precision_list = []
+    recall_list = []
+    meaningful_threshold_list = []
+    for threshold in threshold_list:
```

Comment:
it looks that some of these computations possible to do with sklearn and already implemented classes (doing with the standard APIs is good for understanding and maintenance), namely
1..Compute RocCurve in privacy_risk_score_analysis() as in 
https://github.com/tensorflow/privacy/blob/b208d9deece83bc251d5a7ca6b54612f06885eae/tensorflow_privacy/privacy/membership_inference_attack/membership_inference_attack.py#L74
2.ROCCurve has tpr, fpr, thresholds (ordered). Then here is to get the index of each element of threshold_list in roc_curve.threshould and to get tpr/fpr from roc_curve

Would it work? WDYT?

BTW, does it make sense to return ROC curve and ROC params (like AUC) and to plot charts of RoCCurve

Reply:
I updated the code with metrics.roc_curve function.
I am not sure whether we should return ROC curve and ROC params here, it just makes the privacy risk score analysis very similar to the previous attacks. So I did not change the function further.

Now we compute the privacy risk score for every training and test point, one way to show the result is that we choose a threshold value on privacy risk score (e.g., 0.8), if a training sample's privacy risk score is larger than the threshold, we think it is highly privacy-risky, then we can tell the customers how many risky points we have and further the indices of those risky points.
I will suggest a meeting after the holiday season to discuss more on how we should present the privacy risk score results.